### PR TITLE
fix/issue-#2800: Fix bug with favorites getting deleted when a favori…

### PR DESCRIPTION
**Pull request by groybal**
…ted portlet gets deleted and a user subequently favorites another portlet or unfavorites a portlet

##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included

##### Description of change

See Issue #2800

Updates have been made such that when a user favorites or unfavorites a portlet, the code will first remove any orphaned favorites (favorites for portlets that have been deleted).  What was happening previously was that the UserLayoutManager.saveLayout() method was finding the orphaned favorites and pruning the node "linked-list" at that point, resulting in all favorites coming after in the list to be deleted along with the orphan.

[individual contributor license agreement]: https://github.com/uPortal-Project/uPortal/blob/master/docs/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/uPortal-Project/uPortal/blob/master/docs/CONTRIBUTING.md#commit
[message properties]: https://github.com/uPortal-Project/uPortal/tree/master/uPortal-webapp/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: uPortal-api/uPortal-api-rest/src/test/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServletTest.java [hunk 1]
```diff
@@ -110,7 +259,109 @@ public void testRemoveFavorite() throws IOException {
...
+                favoritedPortlet1Id, MissingPortletDefinition.CHANNEL_ID, favoritedPortlet3Id);
+        ModelAndView mav =
+                updatePreferencesServlet.addFavorite(idForPortletBeingFavorited, req, res);
+        assertNotNull(mav);
+        assertEquals("jsonView", mav.getViewName());
+        assertEquals(idForPortletBeingFavorited, mav.getModel().get("newNodeId"));
+        verify(ulm).deleteNode(orphanedFavoritedPortletSubscribeId);
+        verify(ulm).addNode(any(), eq(favoritesFolderColumnNodeId), eq((String) null));
+        verify(ulm).saveUserLayout();
+    }
```
Comment (ChristianMurphy) [inline uPortal-api/uPortal-api-rest/src/test/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServletTest.java:299]: Would it also make sense to test for an error case like null id?
Something like:

```java
@Test
public void testAddFavoriteWithNullPortletId() throws IOException {
    MockHttpServletRequest request = new MockHttpServletRequest();
    MockHttpServletResponse response = new MockHttpServletResponse();
    
    ModelAndView mav = updatePreferencesServlet.addFavorite(null, request, response);
    
    assertNotNull(mav);
    assertEquals("errorView", mav.getViewName());
    verify(logger).error("Invalid input parameters for adding favorite.");
}
```

**Review by ChristianMurphy — COMMENTED** [review]
Thanks @groybal!
When you have a moment could you look into the failing CI job?

**Comment by groybal** [comment]
> Thanks @groybal! When you have a moment could you look into the failing CI job?

Seems to be working now.

File: uPortal-api/uPortal-api-rest/src/test/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServletTest.java (hunk 1 above)
Comment (groybal) [reply uPortal-api/uPortal-api-rest/src/test/java/org/apereo/portal/layout/dlm/remoting/UpdatePreferencesServletTest.java:297]: Sure, I added a test for null id.

**Review by ChristianMurphy — APPROVED** [review]
